File: current_session/1138.py
"""
see https://leetcode.com/problems/alphabet-board-path/
"""
class Solution(object):
    def alphabetBoardPath(self, target):
        """
        :type target: str
        :rtype: str
        """
        ans, prev = "", "a"
        
        for i in range(len(target)):
            ans += self.getPath(prev, target[i])
            prev = target[i]
        return ans

    # given character x and y, return the path
    def getPath(self, s, e):
        
        if s == e:
            return "!"

        res = ""
        n1, n2 = ord(s)-97, ord(e)-97
        r1, c1 = n1//5, n1%5
        r2, c2 = n2//5, n2%5
        
        # put UL before DR to handle 'z' position
        if r1 > r2:
            res += "U"*(r1-r2)
        if c1 > c2:
            res += "L"*(c1-c2)
        if r1 < r2:
            res += "D"*(r2-r1)
        if c1 < c2:
            res += "R"*(c2-c1)
        return res+"!"

print(Solution().alphabetBoardPath("leet"))
print(Solution().alphabetBoardPath("code"))

File: current_session/c++/1138.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// bfs
// or simply go to the target (?)
// clang++ -std=c++17 c++/1138.cpp -o 1138 && ./1138 -v 
class Solution {
public:
  string alphabetBoardPath(string target) {
    vector<string> res;
    char prev = 'a';
    for (int i = 0; i < target.size(); ++i) {
      string moves = finder(prev, target[i]);
      res.push_back(moves);
      prev = target[i];
    }
    string ans = "";
    for (int i = 0; i < res.size(); ++i) {
      ans += res[i];
    }
    return ans;
  }

private:
  string finder(char prev, char tgt) {
    if (prev == tgt) {
      return "!";
    }
    string moves = "";
    pair<int, int> start = getCoord(prev);
    pair<int, int> end = getCoord(tgt);
    int y1 = start.first;
    int x1 = start.second;
    int y2 = end.first;
    int x2 = end.second;
    // cout << "prev: " << prev << x1 << y1 << endl;
    // cout << "tgt:  " << tgt << x2 << y2 << endl;

    while (y1 > y2) {
      moves.push_back('U');
      y1 -= 1;
    }
    while (x1 > x2) {
      moves.push_back('L');
      x1 -= 1;
    }
    while (y1 < y2) {
      moves.push_back('D');
      y1 += 1;
    }
    while (x1 < x2) {
      moves.push_back('R');
      x1 += 1;
    }
    return moves + "!";
  }

  pair<int, int> getCoord(char x) {
    vector<string> board = {"abcde", "fghij", "klmno", "pqrst", "uvwxy", "z####"};
    for (int i = 0; i < board.size(); ++i) {
      for (int j = 0; j < board[0].size(); ++j) {
        if (board[i][j] == x) {
          return pair<int, int>{i,j};
        }
      }
    }
    cout << "???" << endl;
    return pair<int, int>{0,0};
  }
};

int main() {
  string test1 = "leet";
  string test2 = "code";
  cout << Solution().alphabetBoardPath(test1) << endl;
  cout << Solution().alphabetBoardPath(test2) << endl;
}